android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/PopupWindow.kt

```diff
@@ -8,109 +8,280 @@ import android.graphics.Rect
+/**
+ * Determine whether the container view can display all menu items (without scrolling) within
+ * the available height.
+ *
+ * @return The original container height if the container view can display all menu items
+ * (without scrolling), else calculate the maximum available container height for a scrollable
+ * view with a half menu item.
+ */
+private fun getContainerHeight(
+    recyclerViewHeight: Int,
+    totalNumberOfMenuItems: Int,
+    originalMenuContainerHeight: Int,
+    anchor: View,
+): Int {
+    // Get the total screen display height.
+    val totalHeight = anchor.rootView.measuredHeight
+
+    // Note: We cannot use getWindowVisibleDisplayFrame() as the height is dynamic based on whether
+    // the keyboard is open.
+    // Get any displayed system bars e.g. top status bar, bottom navigation bar or soft buttons bar.
+    val systemBars =
+        ViewCompat.getRootWindowInsets(anchor)?.getInsets(WindowInsetsCompat.Type.systemBars())
+    // Store the vertical status bars.
+    val topSystemBarHeight = systemBars?.top ?: 0
+    val bottomSystemBarHeight = systemBars?.bottom ?: 0
+
+    // Deduct any status bar heights from the total height.
+    val availableHeight = totalHeight - (bottomSystemBarHeight + topSystemBarHeight)
+
+    val menuItemHeight = recyclerViewHeight / totalNumberOfMenuItems
+
+    // We must take the menu container padding into account as this will be applied to the final height.
+    val containerPadding = originalMenuContainerHeight - recyclerViewHeight
+
+    val maxAvailableHeightForRecyclerView = availableHeight - containerPadding
+
+    // The number of menu items that can fit exactly (no cropping) within the max app height.
+    // Round the number of items to the closet Int value to ensure the max space available is utilized.
+    // E.g if 6.9 items fit, round to 7 so the calculation below will show 6.5 items instead of 5.5 .
+    val numberOfItemsFitExactly =
+        (maxAvailableHeightForRecyclerView.toFloat() / menuItemHeight.toFloat()).roundToInt()
+
+    val itemsAlreadyFitContainerHeight = totalNumberOfMenuItems <= numberOfItemsFitExactly
+
+    return if (itemsAlreadyFitContainerHeight) {
+        originalMenuContainerHeight
+    } else {
+        getCroppedMenuContainerHeight(
+            numberOfItemsFitExactly,
+            menuItemHeight,
+            containerPadding,
+        )
+    }
+}
+
+private fun getCroppedMenuContainerHeight(
+    numberOfItemsFitExactly: Int,
+    menuItemHeight: Int,
+    containerPadding: Int,
+): Int {
+    val numberOfItemsFitWithOverflow = numberOfItemsFitExactly.numberOfItemsFitWithOverflow()
+    val updatedRecyclerViewHeight = (numberOfItemsFitWithOverflow * menuItemHeight).toInt()
+
+    return updatedRecyclerViewHeight + containerPadding
+}
+
+/**
+ * The number of menu items that fit exactly, minus a half menu item (indicates more menu items exist).
+ */
+private fun Int.numberOfItemsFitWithOverflow() = this - HALF_MENU_ITEM
```

Comment:
Nit: same comment to a single line function as before, I find it to be an overkill for what it hides :)

 ```
   val numberOfItemsFitWithOverflow = numberOfItemsFitExactly - HALF_MENU_ITEM
```

Reply:
That's fair. It was doing more prior to the latest change, but agree that it is overkill now